framework/src/kernels/ConservativeAdvection.C

```diff
@@ -11,104 +11,175 @@
 #include "SystemBase.h"
 
 registerMooseObject("MooseApp", ConservativeAdvection);
+registerMooseObject("MooseApp", ADConservativeAdvection);
 
+template <bool is_ad>
 InputParameters
-ConservativeAdvection::validParams()
+ConservativeAdvectionTempl<is_ad>::generalParams()
 {
-  InputParameters params = Kernel::validParams();
+  InputParameters params = GenericKernel<is_ad>::validParams();
   params.addClassDescription("Conservative form of $\\nabla \\cdot \\vec{v} u$ which in its weak "
                              "form is given by: $(-\\nabla \\psi_i, \\vec{v} u)$.");
-  params.addRequiredCoupledVar("velocity", "Velocity vector");
   MooseEnum upwinding_type("none full", "none");
   params.addParam<MooseEnum>("upwinding_type",
                              upwinding_type,
                              "Type of upwinding used.  None: Typically results in overshoots and "
                              "undershoots, but numerical diffusion is minimized.  Full: Overshoots "
                              "and undershoots are avoided, but numerical diffusion is large");
+  params.addParam<MaterialPropertyName>("advected_quantity",
+                                        "An optional material property to be advected. If not "
+                                        "supplied, then the variable will be used.");
+  params.addCoupledVar("velocity_variable", "Velocity vector given as a variable");
+  params.addParam<MaterialPropertyName>("velocity_material", "Velocity vector given as a material");
   return params;
 }
 
-ConservativeAdvection::ConservativeAdvection(const InputParameters & parameters)
-  : Kernel(parameters),
-    _velocity(coupledVectorValue("velocity")),
-    _upwinding(getParam<MooseEnum>("upwinding_type").getEnum<UpwindingType>()),
+template <>
+InputParameters
+ConservativeAdvectionTempl<false>::validParams()
+{
+  InputParameters params = generalParams();
+  params.addCoupledVar("velocity", "Velocity vector");
+  params.deprecateCoupledVar("velocity", "velocity_variable ", "12/31/2025");
```

Comment:
The extra space in `velocity_variable` is deliberate so it doesn't screw up the ability to check if `velocity` or `velocity_variable` are valid. I could use `addDeprecatedCoupledVar` with success, but that is a deprecated method.

Reply:
I just don't understand how I'm the first person to hit this problem...

Reply:
fixed it, not doing this hacky way anymore.